invenio_cache/lock.py

```diff
@@ -0,0 +1,195 @@
+            acquired = lock.acquire()
+            if acquired:
+                    f()
+                    lock.release()
+
+        :param lock_id: id of the lock.
+        :type lock_id: str
+        """
+        self.lock_id = lock_id
+        self._acquired = False
+        self._created = datetime.utcnow()
+
+    def __enter__(self):
+        """Entering the context."""
+        return self
+
+    def __exit__(self, exc_type, *args):
+        """Releases the lock if it was previously acquired."""
+        if self.acquired:
+            self.release()
+
+    @property
+    def acquired(self):
+        """Property to check if the lock was acquired."""
+        return self._acquired
+
+    @acquired.setter
+    def acquired(self, value):
+        self._acquired = value
+
+    @property
+    def created(self):
+        """Lock creation date time."""
+        return self._created
+
+    def acquire(self, **kwargs):
+        """Attempts to acquire the lock.
+
+        The method to adcquire the lock must be atomic.
+        It must return a boolean.
+        """
+        raise NotImplementedError
+
+    def release(self):
+        """Attempts to release the lock."""
+        raise NotImplementedError
+
+    def exists(self):
+        """Checks if the lock exists."""
+        raise NotImplementedError
+
+
+class CachedMutex(Lock):
+    """Implements a Mutex using CacheLib API.
+
+    This lock will set a key in a cache when the lock is acquired, and delete the key when
+    the lock is to be deleted.
+    This mechanism is based on the principle that the desired cache's backend provides an atomic
+    write operation (e.g. "write if not exists").
+    """
+
+    _cache = current_cache
+
+    def acquire(self, timeout):
+        """Attempts to acquire the lock.
+
+        The method to adcquire the lock must be atomic.
+
+        :returns: ``True`` if the lock was acquired, ``False`` otherwise .
+        :rtype: boolean
+        :param timeout: lock key timeout.
+        :type timeout: int, optional
+        """
+        success = False
+        try:
+            # Atomic operation to get the lock
+            success = self._cache.add(self.lock_id, True, timeout=timeout)
+        except:
+            # Unexpected error with the cache, we just log it and re-raise
+            current_app.logger.error(
+                f"Unexpected backend failure when acquiring lock {self.lock_id}."
+            )
+            raise
+        self.acquired = success
+        return success
+
+    def release(self):
+        """Attempts to release the lock.
+
+        The lock will only be released if it was previously acquired by this instance.
+
+        :returns: ``True`` if the lock was released, ``False`` otherwise .
+        :rtype: boolean
+        """
+        success = False
+        try:
+            # Release the lock only if it was acquired by this instance
+            if self.acquired:
+                success = self._cache.delete(self.lock_id)
```

Comment:
If the lock was **not** acquired for some reasons, we might want to return `True` or to fail:
```suggestion
            if self.acquired:
                success = self._cache.delete(self.lock_id)
            else:
                success = True
```
Otherwise, it might return False and you don't know what happened.
WDYT?

Reply:
Re-iterating on this: I would remove all `if self.aquired`. You need to be sure to pass the `CachedMutex` instance to release. I would rely exclusively on the existence of the cache key instead.
What is the use case? Multiprocessing context?

My worry is the following: if improperly used, it might happen that I acquire in one process (one web node), and try to release on another process (another web node) later one. This will fail, unless the same instance of `CachedMutex` is shared.

Reply:
I agree with that. 

However, the lock might be released by any thread right now. 

E.g.

```python
# Thread 1
lock = Mutex("id1")
lock.acquire()

# Thread 2
lock = Mutex("id1")
lock.release()
```

We could "sign" these locks with a unique ID (e.g. set the value to a UUID) and only allow the release / renewal if the thread passes the same uuid. We briefly talked about it but did not elaborate on it.